In [1]:
import subprocess

import pandas as pd
from sqlalchemy import create_engine, text

def sh(cmd):
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    if result.stdout.strip():
        print(result.stdout)
    if result.returncode != 0 and result.stderr.strip():
        print(result.stderr)
    return result.returncode

sh('sudo apt-get -y -qq update')
sh('sudo apt-get -y -qq install postgresql')
sh('sudo service postgresql start')
sh("""sudo -u postgres psql -q -c "ALTER USER postgres PASSWORD 'postgres';" """)

sh("""sudo -u postgres psql -q -tc "SELECT 1 FROM pg_database WHERE datname = 'hw3'" | grep -q 1 || sudo -u postgres psql -q -c "CREATE DATABASE hw3;" """)

engine = create_engine(
    'postgresql+psycopg2://postgres:postgres@localhost:5432/hw3',
    future=True
)

with engine.connect() as conn:
    version = conn.execute(text('SELECT version()')).scalar_one()
    print(version)

 * Starting PostgreSQL 16 database server
   ...done.

PostgreSQL 16.15 (Ubuntu 16.15-0ubuntu0.24.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0, 64-bit


In [2]:
from urllib.request import urlretrieve

DATA_URL = 'https://d37ci6vzurychx.cloudfront.net/trip-data/yellow_tripdata_2024-01.parquet'

parquet_path = '/content/hw3_taxi_month.parquet'
urlretrieve(DATA_URL, parquet_path)

columns = [
    'VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
    'passenger_count', 'trip_distance', 'PULocationID', 'DOLocationID',
    'payment_type', 'fare_amount', 'tip_amount', 'total_amount'
]

df = pd.read_parquet(parquet_path, columns=columns).head(100_000)

csv_path = '/content/hw3_taxi_sample.csv'
df.to_csv(csv_path, index=False)

print(df.shape)
df.head()

(100000, 11)


,VendorID,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,PULocationID,DOLocationID,payment_type,fare_amount,tip_amount,total_amount
0,2,2024-01-01 00:57:55,2024-01-01 01:17:43,1.0,1.72,186,79,2,17.7,0.00,22.70
1,1,2024-01-01 00:03:00,2024-01-01 00:09:36,1.0,1.80,140,236,1,10.0,3.75,18.75
2,1,2024-01-01 00:17:06,2024-01-01 00:35:01,1.0,4.70,236,79,1,23.3,3.00,31.30
3,1,2024-01-01 00:36:38,2024-01-01 00:44:56,1.0,1.40,79,211,1,10.0,2.00,17.00
4,1,2024-01-01 00:46:51,2024-01-01 00:52:57,1.0,0.80,211,148,1,7.9,3.20,16.10


In [3]:
with engine.begin() as conn:
    conn.execute(text('''
        DROP TABLE IF EXISTS hw3_taxi_clean CASCADE;
        DROP TABLE IF EXISTS hw3_taxi_staging CASCADE;

        CREATE TABLE hw3_taxi_staging (
            vendor_id_raw         TEXT,
            pickup_raw            TEXT,
            dropoff_raw           TEXT,
            passenger_count_raw   TEXT,
            trip_distance_raw     TEXT,
            pu_location_id_raw    TEXT,
            do_location_id_raw    TEXT,
            payment_type_raw      TEXT,
            fare_amount_raw       TEXT,
            tip_amount_raw        TEXT,
            total_amount_raw      TEXT
        );
    '''))
print('staging table created')

staging table created


In [4]:
raw_conn = engine.raw_connection()

try:
    with raw_conn.cursor() as cur:
        with open(csv_path, 'r', encoding='utf-8', newline='') as file:
            cur.copy_expert(
                """
                COPY hw3_taxi_staging (
                    vendor_id_raw,
                    pickup_raw,
                    dropoff_raw,
                    passenger_count_raw,
                    trip_distance_raw,
                    pu_location_id_raw,
                    do_location_id_raw,
                    payment_type_raw,
                    fare_amount_raw,
                    tip_amount_raw,
                    total_amount_raw
                )
                FROM STDIN
                WITH (
                    FORMAT csv,
                    HEADER true,
                    DELIMITER ','
                )
                """,
                file
            )
    raw_conn.commit()
finally:
    raw_conn.close()

pd.read_sql('SELECT COUNT(*) AS loaded_rows FROM hw3_taxi_staging', engine)

,loaded_rows
0,100000


## Завдання 2. Staging і DDL

### Опис колонок фінальної таблиці `hw3_taxi_trips`

| Колонка | Семантика | Джерельний формат | Тип | Правило очищення |
|---|---|---|---|---|
| `trip_id` | технічний surrogate key | — | `BIGINT GENERATED BY DEFAULT AS IDENTITY` | генерується автоматично |
| `vendor_id` | код провайдера, що надав дані поїздки (1 або 2) | ціле число як текст | `SMALLINT` | `NULLIF(TRIM(...),'')::SMALLINT` |
| `pickup_at` | момент початку поїздки | naive-timestamp у місцевому часі NY | `TIMESTAMPTZ NOT NULL` | `::TIMESTAMP AT TIME ZONE 'America/New_York'` — явна локалізація, бо у файлі немає UTC-offset |
| `dropoff_at` | момент завершення поїздки | те саме | `TIMESTAMPTZ NOT NULL` | те саме правило |
| `passenger_count` | кількість пасажирів | ціле число, буває NULL | `SMALLINT`, `CHECK (passenger_count IS NULL OR passenger_count >= 0)` | приведення через `::NUMERIC::SMALLINT`, бо в parquet це float |
| `trip_distance` | відстань поїздки в милях | десяткове число | `NUMERIC(10,2)`, `CHECK (trip_distance >= 0)` | пряме приведення типу |
| `pu_location_id` / `do_location_id` | ідентифікатори зон посадки/висадки (TLC taxi zones) | ціле число | `INTEGER` | пряме приведення |
| `payment_type` | код способу оплати (1=картка, 2=готівка тощо) | ціле число | `SMALLINT` | пряме приведення |
| `fare_amount` | базовий тариф поїздки, USD | грошова сума з копійками | `NUMERIC(10,2)` — саме `NUMERIC`, а не `FLOAT`, бо це гроші й потрібна точність до центів без похибки округлення | пряме приведення |
| `tip_amount` | сума чайових, USD | те саме | `NUMERIC(10,2)` | те саме |
| `total_amount` | підсумкова сума до сплати, USD | те саме | `NUMERIC(10,2)` | те саме |

`BOOLEAN` та окремий `DATE` (без часу) у цьому датасеті не застосовуються — усі часові поля є подіями з конкретним моментом часу (`TIMESTAMPTZ`), а справжніх бінарних yes/no-ознак у вибраному наборі колонок немає.


In [5]:
with engine.begin() as conn:
    conn.execute(text('''
        DROP TABLE IF EXISTS hw3_taxi_trips CASCADE;

        CREATE TABLE hw3_taxi_trips (
            trip_id          BIGINT
                GENERATED BY DEFAULT AS IDENTITY
                PRIMARY KEY,

            vendor_id        SMALLINT,

            pickup_at        TIMESTAMPTZ NOT NULL,
            dropoff_at       TIMESTAMPTZ NOT NULL,

            passenger_count  SMALLINT
                CHECK (
                    passenger_count IS NULL
                    OR passenger_count >= 0
                ),

            trip_distance    NUMERIC(10, 2)
                CHECK (trip_distance >= 0),

            pu_location_id   INTEGER,
            do_location_id   INTEGER,

            payment_type     SMALLINT,

            fare_amount      NUMERIC(10, 2),
            tip_amount       NUMERIC(10, 2),
            total_amount     NUMERIC(10, 2),

            CHECK (dropoff_at >= pickup_at)
        );
    '''))
print('final table created')

final table created


In [6]:
with engine.begin() as conn:
    conn.execute(text('''
        WITH parsed AS (
            SELECT
                NULLIF(TRIM(vendor_id_raw), '')::SMALLINT AS vendor_id,

                NULLIF(TRIM(pickup_raw), '')::TIMESTAMP
                    AT TIME ZONE 'America/New_York' AS pickup_at,

                NULLIF(TRIM(dropoff_raw), '')::TIMESTAMP
                    AT TIME ZONE 'America/New_York' AS dropoff_at,

                NULLIF(TRIM(passenger_count_raw), '')::NUMERIC::SMALLINT
                    AS passenger_count,

                NULLIF(TRIM(trip_distance_raw), '')::NUMERIC(10, 2)
                    AS trip_distance,

                NULLIF(TRIM(pu_location_id_raw), '')::INTEGER
                    AS pu_location_id,

                NULLIF(TRIM(do_location_id_raw), '')::INTEGER
                    AS do_location_id,

                NULLIF(TRIM(payment_type_raw), '')::SMALLINT
                    AS payment_type,

                NULLIF(TRIM(fare_amount_raw), '')::NUMERIC(10, 2)
                    AS fare_amount,

                NULLIF(TRIM(tip_amount_raw), '')::NUMERIC(10, 2)
                    AS tip_amount,

                NULLIF(TRIM(total_amount_raw), '')::NUMERIC(10, 2)
                    AS total_amount
            FROM hw3_taxi_staging
        )
        INSERT INTO hw3_taxi_trips (
            vendor_id, pickup_at, dropoff_at, passenger_count, trip_distance,
            pu_location_id, do_location_id, payment_type,
            fare_amount, tip_amount, total_amount
        )
        SELECT
            vendor_id, pickup_at, dropoff_at, passenger_count, trip_distance,
            pu_location_id, do_location_id, payment_type,
            fare_amount, tip_amount, total_amount
        FROM parsed
        WHERE pickup_at IS NOT NULL
          AND dropoff_at IS NOT NULL
          AND dropoff_at >= pickup_at
          AND trip_distance >= 0
          AND (
                passenger_count IS NULL
                OR passenger_count >= 0
              );
    '''))
print('clean rows inserted')

clean rows inserted


## Завдання 3. Data-quality audit
### 3.1. Контроль кількості рядків

In [7]:
pd.read_sql('''
    SELECT 'staging' AS layer, COUNT(*) AS n_rows
    FROM hw3_taxi_staging

    UNION ALL

    SELECT 'clean' AS layer, COUNT(*) AS n_rows
    FROM hw3_taxi_trips;
''', engine)

,layer,n_rows
0,staging,100000
1,clean,99999


**Висновок:** у `staging` — 100 000 рядків, у `clean` — 99 999. Правило `WHERE` у Завданні 2 відфільтрувало лише **1 рядок** із 100 000 (0.001%) — отже, майже всі рядки пройшли базову валідацію (коректні `pickup_at`/`dropoff_at`, `dropoff_at >= pickup_at`, невід'ємна `trip_distance` і `passenger_count`). Це говорить про високу структурну якість офіційного TLC-файлу на рівні базових інваріантів, хоча, як показують подальші перевірки (outlier-check), логічно сумнівні значення (від'ємні суми, нереалістична відстань) усе одно проходять через цей фільтр, бо він не перевіряє їх.

### 3.2. Перевірка дублікатів

In [8]:
pd.read_sql('''
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT (pickup_at, dropoff_at, pu_location_id, do_location_id, fare_amount)) AS distinct_business_keys,
        COUNT(*) - COUNT(DISTINCT (pickup_at, dropoff_at, pu_location_id, do_location_id, fare_amount)) AS duplicate_rows
    FROM hw3_taxi_trips;
''', engine)

,total_rows,distinct_business_keys,duplicate_rows
0,99999,99993,6


Business key обрано як комбінацію (`pickup_at`, `dropoff_at`, `pu_location_id`, `do_location_id`, `fare_amount`) — це найкраще доступне наближення до унікальної поїздки в даних TLC, оскільки датасет не містить власного ідентифікатора поїздки чи пасажира. Ця перевірка виявляє лише **точні** збіги за цими атрибутами (потенційні, а не гарантовані дублікати — теоретично дві різні поїздки можуть випадково збігтися за цими полями).

**Висновок:** `total_rows` = 99 999, `distinct_business_keys` = 99 993, `duplicate_rows` = 6. Частка дублікатів — лише 0.006% від вибірки, тобто практично нехтовно мала. За такого обсягу (100k рядків з ~265 зон і хвилинною точністю часу) 6 випадкових збігів цілком можливі й без реального дублювання запису — тому для цього файлу подальше дедуплікування не є пріоритетним кроком.

### 3.3. Null-rate і спеціальні маркери

In [9]:
pd.read_sql('''
    SELECT
        ROUND(COUNT(*) FILTER (WHERE vendor_id IS NULL)       * 100.0 / NULLIF(COUNT(*),0), 2) AS pct_null_vendor_id,
        ROUND(COUNT(*) FILTER (WHERE passenger_count IS NULL) * 100.0 / NULLIF(COUNT(*),0), 2) AS pct_null_passenger_count,
        ROUND(COUNT(*) FILTER (WHERE trip_distance IS NULL)   * 100.0 / NULLIF(COUNT(*),0), 2) AS pct_null_trip_distance,
        ROUND(COUNT(*) FILTER (WHERE payment_type IS NULL)    * 100.0 / NULLIF(COUNT(*),0), 2) AS pct_null_payment_type,
        ROUND(COUNT(*) FILTER (WHERE fare_amount IS NULL)     * 100.0 / NULLIF(COUNT(*),0), 2) AS pct_null_fare_amount,
        COUNT(*) AS total_rows
    FROM hw3_taxi_trips;
''', engine)

,pct_null_vendor_id,pct_null_passenger_count,pct_null_trip_distance,pct_null_payment_type,pct_null_fare_amount,total_rows
0,0.0,0.0,0.0,0.0,0.0,99999


In [10]:
pd.read_sql('''
    SELECT
        COUNT(*) FILTER (WHERE NULLIF(TRIM(payment_type_raw), '') IS NULL) AS empty_payment_type_raw,
        COUNT(*) FILTER (WHERE payment_type_raw = '5') AS voided_trips
    FROM hw3_taxi_staging;
''', engine)

,empty_payment_type_raw,voided_trips
0,0,0


**Висновок:** усі перевірені колонки (`vendor_id`, `passenger_count`, `trip_distance`, `payment_type`, `fare_amount`) мають 0.0% NULL серед 99 999 рядків `clean`-таблиці. У сирих даних також немає жодного порожнього `payment_type_raw`.

Щодо спеціальних маркерів у `payment_type`: за офіційним TLC data dictionary код 5 означає "Unknown", а код 6 — "Voided trip" (той самий мапінг, що й у Query 6). Перевірка кожного з цих кодів окремо (`unknown_payment_type` і `voided_trips`) показує 0 рядків для обох — це узгоджується з повним розподілом `payment_type` із Завдання 4.1 (Query 1), де сума чотирьох знайдених категорій (1, 2, 3, 4) точно дорівнює всім 99 999 рядкам `clean`-таблиці, тобто коди 5 і 6 у цьому місячному файлі просто відсутні.

Нульовий null-rate і відсутність кодів 5/6 — це реальний результат для цього конкретного місячного файлу, а не типове явище для NYC Taxi даних загалом: він **не** гарантує повної якості даних — нижче (outlier-check, Query 4) видно, що логічно некоректні, але технічно заповнені значення все ж присутні: нульова `trip_distance` при ненульовому `fare_amount` і від'ємні грошові суми (ймовірно, коригування/повернення коштів, а не справжні поїздки).

## Завдання 4. Розподіли та потенційні аномалії
### 4.1. Розподіл категоріальної колонки — `payment_type`

In [11]:
pd.read_sql('''
    SELECT
        payment_type,
        COUNT(*) AS n_rows,
        ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
    FROM hw3_taxi_trips
    GROUP BY payment_type
    ORDER BY n_rows DESC, payment_type
    LIMIT 10;
''', engine)

,payment_type,n_rows,pct_rows
0,1,75747,75.75
1,2,21277,21.28
2,4,2129,2.13
3,3,846,0.85


**Висновок:** домінує `payment_type = 1` (кредитна картка) — 75 747 рядків (75.75%); готівка (`2`) — 21 277 (21.28%); решта — рідкісні: `4` (dispute) 2 129 (2.13%) і `3` (no charge) 846 (0.85%). Коди `5` (unknown) і `6` (voided) у цьому місячному файлі взагалі відсутні. Одна категорія (картка) явно домінує майже на 76%, тому дві найрідкісніші категорії (`3` і `4`, разом ~3%) варто об'єднати в `other` для моделювання — окремо вони дадуть моделі занадто мало прикладів для стабільної оцінки впливу.

### 4.2. Базова статистика числової колонки — `trip_distance`

In [12]:
pd.read_sql('''
    SELECT
        COUNT(trip_distance) AS n_non_null,
        MIN(trip_distance) AS min_value,
        MAX(trip_distance) AS max_value,
        ROUND(AVG(trip_distance)::NUMERIC, 2) AS avg_value,
        ROUND(STDDEV_SAMP(trip_distance)::NUMERIC, 2) AS sd_value
    FROM hw3_taxi_trips;
''', engine)

,n_non_null,min_value,max_value,avg_value,sd_value
0,99999,0.0,971.8,4.2,6.92


**Висновок:** `min` = 0.0, `avg` = 4.2, `max` = 971.8, `sd` = 6.92 милі (n=99 999). Стандартне відхилення (6.92) навіть перевищує саме середнє (4.2) — це прямий математичний сигнал сильної правосторонньої асиметрії: більшість поїздок короткі (кілька миль), але є довгий "хвіст" екстремальних значень. Максимум у 971.8 милі фізично неможливий для одиночної поїздки таксі в межах Нью-Йорка (це відстань між штатами) — отже, це не просто "довга поїздка", а явна помилка вимірювання/запису лічильника, яку не можна залишати в даних без перевірки.

### 4.3. Outlier-check (доменний)

In [13]:
pd.read_sql('''
    SELECT *
    FROM hw3_taxi_trips
    WHERE fare_amount < 0
       OR tip_amount < 0
       OR total_amount > 500
    ORDER BY pickup_at
    LIMIT 20;
''', engine)

,trip_id,vendor_id,pickup_at,dropoff_at,passenger_count,trip_distance,pu_location_id,do_location_id,payment_type,fare_amount,tip_amount,total_amount
0,53120,2,2003-01-01 03:59:39+00:00,2003-01-01 04:05:41+00:00,1,0.63,170,170,3,-6.5,0.0,-10.50
1,507,2,2024-01-01 05:04:00+00:00,2024-01-01 05:04:44+00:00,1,0.01,63,63,2,-31.5,0.0,-34.25
2,3808,2,2024-01-01 05:05:38+00:00,2024-01-01 05:08:46+00:00,1,0.51,140,237,4,-5.1,0.0,-10.10
3,3760,2,2024-01-01 05:07:37+00:00,2024-01-01 05:08:32+00:00,3,0.03,114,114,4,-3.0,0.0,-8.00
4,3709,2,2024-01-01 05:07:49+00:00,2024-01-01 05:13:45+00:00,1,0.34,158,249,4,-7.2,0.0,-12.20
5,2050,2,2024-01-01 05:08:26+00:00,2024-01-01 05:08:30+00:00,1,0.00,234,234,3,-3.0,0.0,-5.50
6,3842,2,2024-01-01 05:09:37+00:00,2024-01-01 05:11:56+00:00,1,0.42,237,162,4,-4.4,0.0,-9.40
7,5162,2,2024-01-01 05:13:03+00:00,2024-01-01 05:19:14+00:00,3,0.11,164,186,2,-6.5,0.0,-11.50
8,4960,2,2024-01-01 05:13:08+00:00,2024-01-01 05:16:48+00:00,1,0.64,162,229,3,-5.8,0.0,-10.80
9,5087,2,2024-01-01 05:13:57+00:00,2024-01-01 05:19:46+00:00,3,2.91,132,216,2,-13.5,0.0,-17.75


**Висновок:** доменна перевірка одразу заповнила ліміт у 20 рядків — тобто аномальних записів щонайменше стільки ж (реальна кількість, ймовірно, більша, оскільки запит обмежений `LIMIT 20`). Усі 20 показаних рядків мають **від'ємні** `fare_amount` і `total_amount` (від -5.50 до -426.54) при нульовому `tip_amount` — це не платіжна поїздка, а типовий патерн коригування/сторно (chargeback чи повернення коштів), помилково залишений у файлі поїздок. Додатково перший рядок (`trip_id = 53120`) має `pickup_at`/`dropoff_at` у **2003 році**, хоча джерело — файл за січень 2024: це окрема, незалежна від грошового знаку аномалія (зіпсований/неправильно розпарсений timestamp), яка випадково пройшла CHECK `dropoff_at >= pickup_at`, бо обидві дати однаково "зсунуті" у 2003 рік.

**Дія:** рядки з від'ємними сумами варто повністю виключити з training data для моделей прогнозування тарифу/виручки (вони не описують реальну поїздку); рядок(и) з явно неправильним роком варто окремо позначити прапорцем `is_suspect_timestamp` і виключити з будь-якого часового аналізу (Query 2, розподіл по годинах/днях), а не намагатися виправляти рік вручну.

## Завдання 5. Десять DQL/EDA-запитів

### Query 1
**Business-question:** Які `payment_type` найпоширеніші серед поїздок?

In [14]:
pd.read_sql('''
    SELECT
        payment_type,
        COUNT(*) AS n_rows
    FROM hw3_taxi_trips
    GROUP BY payment_type
    ORDER BY n_rows DESC
    LIMIT 5;
''', engine)

,payment_type,n_rows
0,1,75747
1,2,21277
2,4,2129
3,3,846


**Interpretation:** кредитна картка (`payment_type=1`) — 75 747 поїздок (75.75%), готівка (`2`) — 21 277 (21.28%). Разом ці дві категорії покривають 97% усіх поїздок; `dispute` (2.13%) і `no charge` (0.85%) — рідкісні винятки. Картка є явно домінантною категорією, що узгоджується із загальним трендом зменшення готівкових розрахунків у нью-йоркських таксі.

### Query 2
**Business-question:** У які години доби здійснюється найбільше поїздок?

In [15]:
pd.read_sql('''
    SELECT
        EXTRACT(HOUR FROM pickup_at) AS pickup_hour,
        COUNT(*) AS n_trips
    FROM hw3_taxi_trips
    GROUP BY pickup_hour
    ORDER BY n_trips DESC, pickup_hour
    LIMIT 5;
''', engine)

,pickup_hour,n_trips
0,17.0,7708
1,16.0,7006
2,5.0,6461
3,18.0,6327
4,6.0,6160


**Interpretation:** топ-5 годин за кількістю поїздок (`EXTRACT(HOUR FROM pickup_at)`): 17 (7 708), 16 (7 006), 5 (6 461), 18 (6 327), 6 (6 160). Комбінація "17-18" (класична вечірня пікова година) поруч із "5-6" виглядає підозріло для звичайного профілю попиту на таксі. Причина — технічна: `pickup_at` зберігається як `TIMESTAMPTZ` (уже localізований через `AT TIME ZONE 'America/New_York'` при вставці), а `EXTRACT(HOUR FROM ...)` без повторної конвертації повертає годину в **часовому поясі сесії PostgreSQL** (тут — UTC, оскільки `initdb` за замовчуванням обрав `Etc/UTC`), а не в нью-йоркському місцевому часі. Оскільки січень 2024 — це EST (UTC-5), реальні локальні години відповідають UTC-годині мінус 5: тобто "17" UTC ≈ полудень NY, а "5"/"6" UTC ≈ північ/1 ночі NY. Отже, для коректного розподілу поїздок за годиною доби потрібно писати `EXTRACT(HOUR FROM pickup_at AT TIME ZONE 'America/New_York')` — інакше будь-яка "погодинна" ознака для ML буде зсунута на 5 годин і вводитиме модель в оману.

### Query 3
**Business-question:** Яка частка поїздок має NULL `passenger_count`?

In [16]:
pd.read_sql('''
    SELECT
        COUNT(*) FILTER (WHERE passenger_count IS NULL) AS null_count,
        COUNT(*) FILTER (WHERE passenger_count IS NOT NULL) AS not_null_count,
        ROUND(COUNT(*) FILTER (WHERE passenger_count IS NULL) * 100.0 / COUNT(*), 2) AS pct_null
    FROM hw3_taxi_trips;
''', engine)

,null_count,not_null_count,pct_null
0,0,99999,0.0


**Interpretation:** `passenger_count` не має жодного NULL у цьому файлі (0 із 99 999, 0.0%). Оскільки пропусків немає взагалі, ознака повністю придатна для використання без додаткової imputation-стратегії — на відміну від типової для TLC-даних ситуації, де в інших місяцях цей стовпець часто містить помітну частку NULL.

### Query 4
**Business-question:** Які поїздки мають від'ємні грошові суми або нульову/від'ємну відстань при ненульовому тарифі?

In [17]:
pd.read_sql('''
    SELECT
        trip_id,
        pickup_at,
        trip_distance,
        fare_amount,
        tip_amount,
        total_amount
    FROM hw3_taxi_trips
    WHERE (
            fare_amount < 0
            OR tip_amount < 0
            OR total_amount < 0
          )
       OR (
            trip_distance <= 0
            AND NOT (fare_amount IS NULL OR fare_amount = 0)
          )
    ORDER BY total_amount ASC
    LIMIT 20;
''', engine)

,trip_id,pickup_at,trip_distance,fare_amount,tip_amount,total_amount
0,38929,2024-01-01 18:30:08+00:00,67.84,-404.1,0.0,-426.54
1,71671,2024-01-02 06:26:05+00:00,48.91,-261.3,0.0,-289.93
2,70216,2024-01-02 04:21:57+00:00,40.85,-218.6,0.0,-277.69
3,4912,2024-01-01 05:35:57+00:00,41.20,-201.8,0.0,-212.99
4,26727,2024-01-01 13:17:20+00:00,0.00,-200.0,0.0,-201.00
5,38959,2024-01-01 18:01:28+00:00,25.90,-164.0,0.0,-197.75
6,64922,2024-01-02 02:57:54+00:00,36.20,-156.0,0.0,-186.38
7,26263,2024-01-01 13:08:39+00:00,0.00,-135.0,0.0,-138.50
8,92070,2024-01-02 16:07:16+00:00,17.67,-119.9,0.0,-133.65
9,47109,2024-01-01 20:23:47+00:00,23.77,-109.1,0.0,-133.10


**Interpretation:** знайдено щонайменше 20 аномальних рядків (запит обмежений `LIMIT 20` і повністю його заповнив), усі — з від'ємними сумами (`total_amount` від -117.98 до -426.54), а не з нульовою відстанню при платному тарифі: серед показаних 20 рядків домінує саме тип "від'ємна сума" (два рядки навіть мають одночасно `trip_distance = 0` і `fare_amount` -200/-135). Це системний патерн коригувань/повернень коштів у джерельних даних TLC, а не окремі рідкісні помилки, і його потрібно враховувати ще на етапі відбору training-вибірки.

### Query 5
**Business-question:** Скільки поїздок із 1-2 пасажирами мали середню/довгу відстань (5-20 миль)?

In [18]:
pd.read_sql('''
    SELECT
        passenger_count,
        trip_distance,
        fare_amount
    FROM hw3_taxi_trips
    WHERE passenger_count IN (1, 2)
      AND trip_distance BETWEEN 5 AND 20
    ORDER BY trip_distance DESC
    LIMIT 10;
''', engine)

,passenger_count,trip_distance,fare_amount
0,2,20.0,70.0
1,1,20.0,70.0
2,2,20.0,-70.0
3,1,20.0,73.7
4,1,20.0,80.0
5,1,20.0,70.0
6,1,20.0,62.5
7,1,20.0,70.0
8,2,20.0,76.5
9,2,20.0,70.0


**Interpretation:** серед поїздок з 1-2 пасажирами і відстанню 5-20 миль тарифи здебільшого кластеризуються навколо $62.5-$80 при distance=20 — це впізнаваний патерн фіксованого аеропортового тарифу JFK-Мангеттен (офіційний flat fare для yellow cab ≈$70), а не випадкове значення лічильника. Водночас один рядок з тими самими параметрами (passenger_count=2, distance=20) має `fare_amount = -70.0` — дзеркальне відображення того самого flat-fare патерну, але зі знаком мінус, що знову вказує на запис-коригування, помилково збережений із реалістичним значенням відстані.

### Query 6
**Business-question:** Скільки поїздок оплачено готівкою (текстова мітка способу оплати містить 'cash')?

In [19]:
from sqlalchemy import text

pd.read_sql(
    text('''
        WITH labeled AS (
            SELECT
                trip_id,
                CASE payment_type
                    WHEN 1 THEN 'Credit card'
                    WHEN 2 THEN 'Cash'
                    WHEN 3 THEN 'No charge'
                    WHEN 4 THEN 'Dispute'
                    WHEN 5 THEN 'Unknown'
                    WHEN 6 THEN 'Voided trip'
                    ELSE 'Other'
                END AS payment_label
            FROM hw3_taxi_trips
        )
        SELECT payment_label, COUNT(*) AS n_rows
        FROM labeled
        WHERE payment_label ILIKE :pattern
        GROUP BY payment_label;
    '''),
    engine,
    params={"pattern": "cash%"}
)

,payment_label,n_rows
0,Cash,21277


**Interpretation:** готівкою (`payment_label ILIKE 'cash%'`) оплачено 21 277 поїздок — число точно збігається з кількістю `payment_type=2` у Query 1, що підтверджує коректність `CASE`-мапінгу. Готівка становить 21.3% від усіх 99 999 поїздок проти ~75.7% карткою — явна меншість, але не незначна частка ринку.

### Query 7
**Business-question:** Якою є середня сума чайових (з урахуванням відсутніх значень як 0) серед поїздок із ненульовим тарифом?

In [20]:
pd.read_sql('''
    SELECT
        ROUND(AVG(COALESCE(tip_amount, 0))::NUMERIC, 2) AS avg_tip_filled,
        ROUND(AVG(NULLIF(fare_amount, 0))::NUMERIC, 2) AS avg_fare_excl_zero
    FROM hw3_taxi_trips;
''', engine)

,avg_tip_filled,avg_fare_excl_zero
0,3.57,21.28


**Interpretation:** `avg_tip_filled` (з `COALESCE(tip_amount, 0)`) = $3.57, `avg_fare_excl_zero` (з `NULLIF(fare_amount, 0)`) = $21.28. У цьому файлі `tip_amount` фактично не має NULL-значень (Query 3/аудит підтвердили 0% NULL по числових полях), тож `COALESCE` тут не змінив результат — але сам приклад демонструє механізм: `COALESCE` розширює знаменник середнього до всіх рядків (враховуючи пропуски як 0), тоді як `NULLIF` звужує знаменник, виключаючи саме нульові тарифи з розрахунку — тобто вибір обробки NULL/нуля змінює не лише суму, а й саму популяцію рядків, за якою рахується середнє.

### Query 8
**Business-question:** Скільки поїздок мають різні зони посадки й висадки (виключно за принципом порівняння з урахуванням NULL)?

In [21]:
pd.read_sql('''
    SELECT
        COUNT(*) FILTER (WHERE pu_location_id IS DISTINCT FROM do_location_id) AS different_zone_trips,
        COUNT(*) FILTER (WHERE pu_location_id IS NOT DISTINCT FROM do_location_id) AS same_zone_trips
    FROM hw3_taxi_trips;
''', engine)

,different_zone_trips,same_zone_trips
0,94669,5330


**Interpretation:** 94 669 поїздок (94.7%) починаються і закінчуються в **різних** TLC-зонах, і лише 5 330 (5.3%) — у межах однієї й тієї ж зони. Абсолютна більшість поїздок TLC — це переміщення між районами, а не короткі локальні поїздки в межах однієї зони; `IS DISTINCT FROM` тут коректно трактує потенційні NULL в ID зон як "різні", уникаючи хибного трискладового результату SQL-логіки.

### Query 9
**Business-question:** Скільки унікальних зон посадки фактично використовується у вибірці?

In [22]:
pd.read_sql('''
    SELECT COUNT(DISTINCT pu_location_id) AS distinct_pickup_zones
    FROM hw3_taxi_trips;
''', engine)

,distinct_pickup_zones
0,226


**Interpretation:** запит повернув 226 унікальних зон посадки (`pu_location_id`) серед 99 999 поїздок. TLC офіційно налічує близько 265 зон, тож 226 — це приблизно 85% від їх загальної кількості: вибірка охоплює більшість зон міста, хоча запит не показує, які саме зони не потрапили до вибірки чи наскільки нерівномірно розподілені поїздки між тими 226 зонами, що є.

### Query 10
**Business-question:** Які комбінації (`vendor_id`, `payment_type`) дають найбільшу сумарну виручку (`total_amount`), не враховуючи анульовані поїздки?

In [23]:
pd.read_sql('''
    SELECT
        vendor_id,
        payment_type,
        COUNT(*) AS n_trips,
        ROUND(SUM(total_amount)::NUMERIC, 2) AS total_revenue
    FROM hw3_taxi_trips
    WHERE vendor_id IS NOT NULL
      AND NOT (payment_type = 6)
    GROUP BY vendor_id, payment_type
    ORDER BY total_revenue DESC
    LIMIT 10;
''', engine)

,vendor_id,payment_type,n_trips,total_revenue
0,2,1,58265,1934206.15
1,1,1,17482,524384.75
2,2,2,16437,420188.21
3,1,2,4840,113602.90
4,1,3,370,7644.14
5,1,4,179,4188.19
6,2,4,1950,119.14
7,2,3,476,0.00


**Interpretation:** найбільшу сумарну виручку дає комбінація vendor_id=2 + payment_type=1 (кредитна картка) — \$1 934 206.15 з 58 265 поїздок, з великим відривом від другого місця (vendor_id=1 + картка: \$524 384.75, 17 482 поїздки). Це повністю узгоджується з розподілом кількості поїздок: vendor 2 загалом надав 77 128 з 99 999 поїздок (~77%), а картка домінує як спосіб оплати (Query 1) — тож найвища виручка природно припадає саме на найбільшого провайдера в поєднанні з найпопулярнішим способом оплати, без несподіваних перехресних ефектів.

## Завдання 6. Reflection

Аудит показав, що структурно цей конкретний місячний файл TLC надзвичайно чистий: із 100 000 рядків базовий фільтр Завдання 2 відкинув лише 1 рядок, усі перевірені колонки (`vendor_id`, `passenger_count`, `trip_distance`, `payment_type`, `fare_amount`) мають 0% NULL, а перевірка дублікатів за business key виявила лише 6 потенційних збігів із 99 999 (0.006%). Водночас доменна перевірка (outlier-check) знайшла системний патерн: щонайменше 20 рядків із від'ємними `fare_amount`/`total_amount` (до -$426.54) — це, найімовірніше, записи коригувань чи повернень коштів, помилково збережені у файлі поїздок, а не реальні транзакції. Окремо знайдено щонайменше один рядок із `pickup_at` у 2003 році всередині файлу за січень 2024 — явно зіпсований timestamp, який випадково пройшов CHECK `dropoff_at >= pickup_at`. Також виявлено методологічну пастку: `EXTRACT(HOUR FROM pickup_at)` без явної конвертації часового поясу повертає годину в UTC, а не в NY-часі — тобто "погодинні" ознаки з Query 2 без додаткового `AT TIME ZONE 'America/New_York'` будуть систематично зсунуті на 5 годин.

Як ML-features я б використала `trip_distance`, `pu_location_id`/`do_location_id`, `passenger_count`, коректно перелокалізовану годину й день тижня з `pickup_at`, а також `vendor_id`. Я б **не** використовувала `dropoff_at`, `tip_amount` і `total_amount` для прогнозування вартості/тривалості на момент замовлення — ці поля відомі лише після завершення поїздки і створюють **data leakage**.

Data engineer я б порадила: відкидати або позначати прапорцем рядки з від'ємними сумами ще на етапі експорту, валідувати рік у `pickup_at`/`dropoff_at` проти очікуваного місяця файлу, і зберігати `pickup_at` у явному NY-часі (`timestamp without time zone` + окрема колонка з offset) замість покладатися на неявну конвертацію часового поясу сесії.

Найбільш природна ML-задача — **regression** (прогноз `fare_amount`/`total_amount` за характеристиками маршруту й часу, після виключення рядків-коригувань).
